File: targets/widget/Snapshot.swift
import Foundation

/// What the widgets can know.
///
/// ⚠⚠ **`widget/snapshot.json` in the App Group is the primary input** (plus,
/// since ADR 0080, `widget/live.json` and one rationed poll — see `Live.swift`;
/// nothing else). A widget
/// extension is a separate process: it cannot call the API, read AsyncStorage,
/// reach the TanStack Query cache (which is not persisted to disk), or import
/// `@/constants/theme` or `@/lib/i18n`. Anything drawn here was put in the
/// container by `src/features/widgets/snapshot.ts` before this ran.
///
/// ⚠⚠ **Every string the widget PRINTS is pre-formatted by JS, including the
/// furniture.** That inverts what the notification content extension does with
/// its two `.lproj` files, and the inversion is the point: a widget renders in
/// the SYSTEM language, so `.lproj` strings would hand an English widget to a
/// reader who picked Spanish inside the app. The snapshot travels with the
/// reader's `lang`, `tz` and `clock` already applied. The only hardcoded English
/// in this target is `placeholder`, for the gallery preview before the app has
/// ever run.
///
/// ⚠ Decoding is forgiving on purpose — this parses a file written by a build of
/// the app that may be older than this extension. A missing field is a quieter
/// widget, never a crash.
struct WidgetSnapshot: Codable {
  struct Club: Codable, Hashable {
    let slug: String
    let name: String
    let abbr: String
  }

  /// The card's own furniture. See the type header for why this is not `.lproj`.
  struct Copy: Codable {
    let next: String
    let yourWeek: String
    /// ⚠ Fully formed — `3 CLUBS` / `3 CLUBES`. The plural is resolved in JS, so
    /// this target holds no counting rule for any language.
    let clubCount: String
    let followPrompt: String
    let noFixtures: String
    /// The separator in `VAL v RMA`. Localised — `v` in English, `vs` in Spanish.
    let versus: String
    /// Prefix for an away row: `at Girona` / `en Girona`.
    let away: String
    /// The pill after the followed side on the medium widget: `HOME` / `CASA`
    /// (ADR 0059). ⚠ Optional — absent from a v1 snapshot, and a missing pill
    /// is the honest degradation.
    let homeTag: String?
    let awayTag: String?
    /// The live ledger's states (ADR 0080): `LIVE` / `EN JUEGO`, `HT` / `DESC`,
    /// `FT` / `FIN`. ⚠ Optional — absent from a v2 snapshot; the English
    /// fallbacks below are the degradation, matching `fallbackCopy`'s rule.
    let live: String?
    let halfTime: String?
    let fullTime: String?

    var liveLabel: String { live ?? "LIVE" }
    var halfTimeLabel: String { halfTime ?? "HT" }
    var fullTimeLabel: String { fullTime ?? "FT" }
  }

  struct Entry: Codable, Identifiable {
    let fixtureId: String
    /// The FOLLOWED club this row belongs to — not necessarily the home side.
    ///
    /// ⚠⚠ **Never the whole answer to "is this club X's row" — use `involves`.**
    /// A derby between two followed clubs is ONE entry (the writer dedupes by
    /// fixture id) and this names only the side `upcomingRow` resolved, home
    /// winning the tie (ADR 0029). Filtering `clubSlug ==` alone made a widget
    /// configured to the derby's away side say "no matches scheduled" while its
    /// club sat right there in the snapshot (ADR 0102).
    let clubSlug: String
    let isHome: Bool
    let homeAbbr: String
    let awayAbbr: String
    /// Both sides' slugs, v4 (ADR 0102) — what `involves` matches. ⚠ Optional:
    /// absent from a v3 snapshot, and absence degrades the club filter to the
    /// old owner-only match, never to a crash.
    let homeSlug: String?
    let awaySlug: String?
    /// Both sides named, home first (ADR 0059). ⚠ Optional: a v1 snapshot
    /// written by an older app has neither, and the row falls back to the abbrs.
    let homeName: String?
    let awayName: String?
    let opponentName: String
    let opponentAbbr: String
    /// Which crest file in the App Group is the opponent's: `home` or `away`.
    let opponentSlot: String
    let kickoffUtc: Date
    /// Pre-formatted in the reader's zone and clock preference. `Sat 21:00`.
    let kickoffLabel: String
    /// `SAT` over `21:00` — the medium widget's stacked column (ADR 0059).
    /// ⚠ Optional for the same reason as `homeName`; the row falls back to
    /// `kickoffLabel` on one line.
    let kickoffDay: String?
    let kickoffTime: String?
    /// `J4`. Null for a competition without rounds.
    let roundLabel: String?
    let venue: String?
    /// The **API** league slug (`laliga`), v3 (ADR 0080) — the poll gate: only
    /// LaLiga goes live on `/cronogol/live`. ⚠ Optional: absent from a v2
    /// snapshot, and absence means "assume eligible", never "skip".
    ///
    /// ⚠⚠ **`laliga`, NOT our own `la-liga`** (ADR 0084). This field is written
    /// straight off `WindowFixtureView.leagueSlug` in
    /// `src/features/widgets/snapshot.ts`, which is the value
    /// `GET /cronogol/fixtures` serves — the same slug `LiveFetch` puts in
    /// `?league=`. The first cut of 0080 compared against `la-liga`, the app's
    /// INTERNAL slug from `src/lib/cronogol/leagues.ts`, so `liveEligible` was
    /// false for every real fixture and the whole live path was dead. See 0084
    /// before "correcting" this back.
    let leagueSlug: String?

    var id: String { fixtureId }

    /// Whether this row is `club`'s match — either side of it, not just the
    /// side `clubSlug` resolved to. See `clubSlug` for why the difference is a
    /// shipped bug and not a nicety.
    func involves(_ club: String) -> Bool {
      clubSlug == club || homeSlug == club || awaySlug == club
    }

    /// Whether `/cronogol/live` could ever serve this fixture.
    ///
    /// ⚠ The comparand is the API slug — see `leagueSlug` above and ADR 0084.
    var liveEligible: Bool { leagueSlug == nil || leagueSlug == "laliga" }
  }

  let v: Int
  let writtenAt: Date
  let clubs: [Club]
  let copy: Copy
  let entries: [Entry]
}

extension WidgetSnapshot {
  /// ⚠ Matches `ios.entitlements` in `app.json`, every `expo-target.config.js`,
  /// and `APP_GROUP` in `src/features/push/crest-cache.ts`.
  static let appGroup = "group.com.altagamafc.app"

  static func load() -> WidgetSnapshot? {
    guard
      let container = FileManager.default.containerURL(
        forSecurityApplicationGroupIdentifier: appGroup
      ),
      let data = try? Data(contentsOf: container.appendingPathComponent("widget/snapshot.json"))
    else { return nil }

    let decoder = JSONDecoder()
    // ⚠ The API emits both `…Z` and `…+00:00`, and only the fractional-seconds
    // formatter accepts the ones carrying milliseconds. Trying one shape only
    // loses every time — the same trap `Payload.swift` documents.
    decoder.dateDecodingStrategy = .custom { decoder in
      let text = try decoder.singleValueContainer().decode(String.self)
      let plain = ISO8601DateFormatter()
      plain.formatOptions = [.withInternetDateTime]
      if let date = plain.date(from: text) { return date }

      let fractional = ISO8601DateFormatter()
      fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
      guard let date = fractional.date(from: text) else {
        throw DecodingError.dataCorruptedError(
          in: try decoder.singleValueContainer(),
          debugDescription: "Not an ISO-8601 instant: \(text)"
        )
      }
      return date
    }

    return try? decoder.decode(WidgetSnapshot.self, from: data)
  }

  /// Rows still ahead of `date`, optionally narrowed to one club.
  ///
  /// ⚠ The filter on `kickoffUtc > date` is what makes a snapshot the app has
  /// not refreshed degrade to the empty state rather than to a countdown that
  /// has gone negative. A widget on a phone that has not opened the app in a
  /// week must say "nothing scheduled", not lie about a match played on Sunday.
  func rows(after date: Date, clubSlug: String?) -> [Entry] {
    entries
      .filter { $0.kickoffUtc > date }
      .filter { entry in clubSlug.map(entry.involves) ?? true }
      .sorted { $0.kickoffUtc < $1.kickoffUtc }
  }

  /// The rows a live ledger could be drawn for at `date` (ADR 0080): kickoff
  /// has passed, and either play could still be running — inside the same
  /// 150-minute hold the Today board uses (`KICKOFF_HOLD_MS`, ADR 0078) — or
  /// the match kicked off earlier the SAME LOCAL DAY, which is how a `finished`
  /// entry in `live.json` keeps its full-time card until midnight.
  ///
  /// ⚠ Live-eligible leagues only. A Premier League fixture in this window has
  /// no live route behind it, and a ledger of dashes that never resolves is
  /// worse than the old behaviour (the row simply dropping at kickoff).
  ///
  /// Earliest kickoff first — the small widget takes `.first`.
  func rows(inPlayAt date: Date, clubSlug: String?, calendar: Calendar = .current) -> [Entry] {
    entries
      .filter { $0.kickoffUtc <= date && $0.liveEligible }
      .filter {
        date.timeIntervalSince($0.kickoffUtc) <= 150 * 60
          || calendar.isDate($0.kickoffUtc, inSameDayAs: date)
      }
      .filter { entry in clubSlug.map(entry.involves) ?? true }
      .sorted { $0.kickoffUtc < $1.kickoffUtc }
  }

  /// Whether the reader follows anything at all — which is a different empty
  /// state from "follows clubs, none of them are playing".
  var followsNothing: Bool { clubs.isEmpty }
}

extension WidgetSnapshot {
  /// ⚠ The only hardcoded copy in this target, and English-only by necessity —
  /// before the app has run there is no reader preference to read. Kept to the
  /// widget's furniture.
  static let fallbackCopy = Copy(
    next: "NEXT",
    yourWeek: "YOUR WEEK",
    clubCount: "",
    followPrompt: "Open Alta Gama FC",
    noFixtures: "No matches scheduled",
    versus: "v",
    away: "at",
    homeTag: "HOME",
    awayTag: "AWAY",
    live: "LIVE",
    halfTime: "HT",
    fullTime: "FT"
  )

  /// No snapshot on disk at all — the app has never run, or the App Group is not
  /// provisioned.
  ///
  /// ⚠⚠ **This is NOT `placeholder`, and the difference is the whole point.** A
  /// widget placed on a real home screen before the app's first launch used to
  /// fall through to the gallery sample and draw `VAL v RMA · Mestalla` — three
  /// fabricated fixtures for clubs the reader may not follow, in a country they
  /// may not live in, presented exactly like real data. The sample belongs in
  /// the picker, where everything is obviously a preview, and nowhere else.
  static func unavailable(relativeTo now: Date) -> WidgetSnapshot {
    WidgetSnapshot(v: 4, writtenAt: now, clubs: [], copy: fallbackCopy, entries: [])
  }

  /// The gallery preview.
  ///
  /// ⚠ For `context.isPreview` only — see `unavailable` for why. A real but empty
  /// snapshot would make every tile in the widget picker say "Follow a club",
  /// which reads as broken to someone who simply has not opened the app yet.
  static func placeholder(relativeTo now: Date) -> WidgetSnapshot {
    WidgetSnapshot(
      v: 4,
      writtenAt: now,
      clubs: [
        .init(slug: "valencia", name: "Valencia CF", abbr: "VAL"),
        .init(slug: "sevilla", name: "Sevilla FC", abbr: "SEV"),
        .init(slug: "athletic", name: "Athletic Club", abbr: "ATH"),
      ],
      copy: .init(
        next: "NEXT",
        yourWeek: "YOUR WEEK",
        clubCount: "3 CLUBS",
        followPrompt: "Follow a club",
        noFixtures: "No fixtures scheduled",
        versus: "v",
        away: "at",
        homeTag: "HOME",
        awayTag: "AWAY",
        live: "LIVE",
        halfTime: "HT",
        fullTime: "FT"
      ),
      entries: [
        .init(
          fixtureId: "preview-1", clubSlug: "valencia", isHome: true,
          homeAbbr: "VAL", awayAbbr: "RMA",
          homeSlug: "valencia", awaySlug: "real-madrid",
          homeName: "Valencia", awayName: "R. Madrid",
          opponentName: "Real Madrid", opponentAbbr: "RMA", opponentSlot: "away",
          kickoffUtc: now.addingTimeInterval(100_800),
          kickoffLabel: "Sat 21:00", kickoffDay: "SAT", kickoffTime: "21:00",
          roundLabel: "J4", venue: "Mestalla", leagueSlug: "laliga"
        ),
        .init(
          fixtureId: "preview-2", clubSlug: "sevilla", isHome: true,
          homeAbbr: "SEV", awayAbbr: "OSA",
          homeSlug: "sevilla", awaySlug: "osasuna",
          homeName: "Sevilla", awayName: "Osasuna",
          opponentName: "Osasuna", opponentAbbr: "OSA", opponentSlot: "away",
          kickoffUtc: now.addingTimeInterval(169_200),
          kickoffLabel: "Sun 16:15", kickoffDay: "SUN", kickoffTime: "16:15",
          roundLabel: "J4", venue: "Sánchez-Pizjuán", leagueSlug: "laliga"
        ),
        .init(
          fixtureId: "preview-3", clubSlug: "athletic", isHome: false,
          homeAbbr: "GIR", awayAbbr: "ATH",
          homeSlug: "girona", awaySlug: "athletic",
          homeName: "Girona", awayName: "Athletic",
          opponentName: "Girona", opponentAbbr: "GIR", opponentSlot: "home",
          kickoffUtc: now.addingTimeInterval(177_300),
          kickoffLabel: "Sun 18:30", kickoffDay: "SUN", kickoffTime: "18:30",
          roundLabel: "J4", venue: "Montilivi", leagueSlug: "laliga"
        ),
      ]
    )
  }
}

File: src/features/widgets/snapshot.ts
/**
 * The widget snapshot — what the WidgetKit extension is allowed to know.
 *
 * ⚠⚠ **A widget extension is a separate process with NO access to this app.** It
 * cannot call the API, read AsyncStorage, reach the TanStack Query cache (which
 * is never persisted to disk), or import `@/constants/theme` or `@/lib/i18n`.
 * Everything it draws has to be sitting in the App Group before it runs. This
 * file is the whole contract; `targets/widget/Snapshot.swift` is its mirror, and
 * a field added to one is a blank row on the card until it is added to both.
 *
 * ⚠⚠ **Every string the widget PRINTS is pre-formatted here, including the
 * furniture.** That inverts what the notification content extension does with
 * its two `.lproj` files, and the inversion is the point (ADR 0047): a widget
 * renders in the SYSTEM language, so `.lproj` strings would hand an English
 * widget to a reader who chose Spanish inside the app — and the same argument
 * covers `tz` and the 12/24h clock, neither of which the extension can see.
 * Only `kickoffUtc` travels raw, because the countdown is arithmetic.
 *
 * ⚠ The selection is PURE and the writing is not, deliberately — the same split
 * `selectReminders`/`applyReminders` already uses, so the whole contract is
 * exercisable from `_debug/widgets` with no native module in sight.
 */
import { Directory, File } from 'expo-file-system';

import { groupContainer } from '@/features/app-group';
import { involvesFollowed, upcomingRow } from '@/lib/cronogol/board';
import { abbreviate, matchday, widgetName } from '@/lib/cronogol/derive';
import type { WindowFixtureView } from '@/lib/cronogol/types';
import type { Copy } from '@/lib/i18n/copy';

/** Bump when `WidgetEntry`/`WidgetSnapshot` changes shape. Swift tolerates both. */
export const SNAPSHOT_VERSION = 2;

/**
 * How many fixtures travel.
 *
 * ⚠ Six, not three, and the extra three are not slack. The small widget needs 1
 * and the medium 3, but the app may not run for days while the widget keeps
 * rendering from a timeline — and the extension drops entries whose kickoff has
 * passed, so a snapshot of exactly 3 degrades to 2 rows, then 1, then empty.
 * Six keeps the medium honest across roughly a fortnight of app silence, and
 * costs at most twelve crest files, most of them already on disk from the
 * reminder path.
 */
export const WIDGET_ENTRY_BUDGET = 6;

/** ⚠ Under the route's 31-day span cap, which is a 400 and not a clamp. */
export const WIDGET_WINDOW_DAYS = 21;

export interface WidgetClub {
  slug: string;
  name: string;
  abbr: string;
}

export interface WidgetEntry {
  fixtureId: string;
  /** The FOLLOWED club, whichever side it is on. Drives the deep link. */
  clubSlug: string;
  isHome: boolean;
  /** ⚠ Home-first, the way the match is named — never "your club" first. */
  homeAbbr: string;
  awayAbbr: string;
  /**
   * Both sides named via `widgetName` (`R. Madrid`, `Athletic`) for the medium
   * widget's row (ADR 0059) — contracted, because two names and a pill share
   * one row and the full forms truncated on the simulator. ⚠ Optional in `Snapshot.swift` — a v1
   * file has neither, and the extension falls back to the abbrs.
   */
  homeName: string;
  awayName: string;
  opponentName: string;
  opponentAbbr: string;
  /** Which crest file in the App Group is the opponent's. */
  opponentSlot: 'home' | 'away';
  /** ⚠ The only raw instant. Everything else is a finished string. */
  kickoffUtc: string;
  kickoffLabel: string;
  /** `SAT` over `21:00` — the medium widget's stacked column (ADR 0059). */
  kickoffDay: string;
  kickoffTime: string;
  roundLabel: string | null;
  venue: string | null;
}

export interface WidgetSnapshot {
  v: number;
  writtenAt: string;
  clubs: WidgetClub[];
  copy: {
    next: string;
    yourWeek: string;
    clubCount: string;
    followPrompt: string;
    noFixtures: string;
    versus: string;
    away: string;
    homeTag: string;
    awayTag: string;
  };
  entries: WidgetEntry[];
}

/**
 * The next match for each followed club, soonest first.
 *
 * ⚠ The eligibility rules are the ones `selectReminders` documents at length and
 * they apply here for the same reasons: a `kickoffTbd` fixture is a date wearing
 * a midnight time, so a countdown to it is a lie, and a `postponed` or
 * `cancelled` match is not a match.
 *
 * ⚠⚠ **Deduped by fixture id AFTER the per-club pass.** A derby between two
 * followed clubs is ONE match, and grouping by club alone puts it on the medium
 * widget twice — two rows, same crest pair, same kickoff. This is the single
 * most likely wrong-looking output of this function.
 */
export function selectWidgetFixtures(
  fixtures: readonly WindowFixtureView[],
  followed: readonly string[],
  now: Date,
  budget: number = WIDGET_ENTRY_BUDGET,
): WindowFixtureView[] {
  const eligible = fixtures
    .filter((fixture) => {
      if (fixture.kickoffTbd) return false;
      if (fixture.status === 'cancelled' || fixture.status === 'postponed') return false;

      const kickoff = Date.parse(fixture.kickoffUtc);
      if (Number.isNaN(kickoff) || kickoff <= now.getTime()) return false;

      return involvesFollowed(fixture, followed);
    })
    .sort((a, b) => Date.parse(a.kickoffUtc) - Date.parse(b.kickoffUtc));

  const seenClubs = new Set<string>();
  const seenFixtures = new Set<string>();
  const picked: WindowFixtureView[] = [];

  for (const fixture of eligible) {
    // ⚠ Both sides are checked, not just the one `upcomingRow` resolves. A match
    // is "the next one" for the home club AND the away club when both are
    // followed; taking only the resolved side would let the loser of that tie
    // skip its own next match and show a later one instead.
    const sides = [fixture.homeTeam?.slug, fixture.awayTeam?.slug]
      .filter((slug): slug is string => typeof slug === 'string' && followed.includes(slug));

    const fresh = sides.filter((slug) => !seenClubs.has(slug));
    if (fresh.length === 0) continue;

    for (const slug of fresh) seenClubs.add(slug);
    if (seenFixtures.has(fixture.id)) continue;
    seenFixtures.add(fixture.id);
    picked.push(fixture);
  }

  return picked.slice(0, budget);
}

/**
 * Everything the widget needs, rendered for THIS reader.
 *
 * ⚠ Pure. `format` and `now` are arguments rather than store reads for the same
 * reason `selectReminders` takes them — it is what keeps the whole thing
 * exercisable without a device.
 */
export function buildSnapshot(
  fixtures: readonly WindowFixtureView[],
  followed: readonly string[],
  now: Date,
  copy: Copy,
  formatKickoff: (iso: string) => string,
  formatKickoffParts: (iso: string) => { day: string; time: string },
): WidgetSnapshot {
  const picked = selectWidgetFixtures(fixtures, followed, now);

  const entries: WidgetEntry[] = [];
  const clubs = new Map<string, WidgetClub>();

  for (const fixture of picked) {
    // ⚠ `upcomingRow` owns "home wins the tie when both sides are followed"
    // (ADR 0029). Re-deriving it here is how the widget and the Today board
    // start disagreeing about whose match a derby is.
    const row = upcomingRow(fixture, followed);
    if (!row || !row.opponent) continue;

    const club = {
      slug: row.club.slug,
      name: row.club.name,
      // ⚠ `shortName` is nullable and NOT unique — Valencia and Valladolid have
      // collided. `abbreviate` is the same guard the club pages use.
      abbr: abbreviate(row.club.name, row.club.slug, row.club.shortName),
    };
    if (!clubs.has(club.slug)) clubs.set(club.slug, club);

    const md = matchday(fixture.round) ?? fixture.matchweek;
    const parts = formatKickoffParts(fixture.kickoffUtc);

    entries.push({
      fixtureId: fixture.id,
      clubSlug: row.club.slug,
      isHome: row.isHome,
      homeAbbr: fixture.homeTeam
        ? abbreviate(fixture.homeTeam.name, fixture.homeTeam.slug, fixture.homeTeam.shortName)
        : '',
      awayAbbr: fixture.awayTeam
        ? abbreviate(fixture.awayTeam.name, fixture.awayTeam.slug, fixture.awayTeam.shortName)
        : '',
      homeName: fixture.homeTeam ? widgetName(fixture.homeTeam.name) : '',
      awayName: fixture.awayTeam ? widgetName(fixture.awayTeam.name) : '',
      opponentName: row.opponent.name,
      opponentAbbr: abbreviate(
        row.opponent.name,
        row.opponent.slug,
        row.opponent.shortName,
      ),
      opponentSlot: row.isHome ? 'away' : 'home',
      kickoffUtc: fixture.kickoffUtc,
      kickoffLabel: formatKickoff(fixture.kickoffUtc),
      kickoffDay: parts.day,
      kickoffTime: parts.time,
      // ⚠ `copy.today.md` — the SAME matchday label the last-result card uses
      // (`J 4` / `MD 4`). The mock draws `J4` closed up; one source of truth for
      // the label is worth more than one space.
      roundLabel: md !== null ? copy.today.md(md) : null,
      venue: fixture.venue,
    });
  }

  // ⚠ The club list is what the Edit Widget picker offers, so it must be every
  // club the reader FOLLOWS — not only the ones with a fixture in the window. A
  // club on an international break would otherwise vanish from the picker.
  for (const slug of followed) {
    if (clubs.has(slug)) continue;
    const found = fixtures
      .flatMap((fixture) => [fixture.homeTeam, fixture.awayTeam])
      .find((team) => team?.slug === slug);
    if (!found) continue;
    clubs.set(slug, {
      slug,
      name: found.name,
      abbr: abbreviate(found.name, found.slug, found.shortName),
    });
  }

  return {
    v: SNAPSHOT_VERSION,
    writtenAt: now.toISOString(),
    clubs: [...clubs.values()],
    copy: {
      next: copy.widgets.next,
      yourWeek: copy.widgets.yourWeek,
      clubCount: copy.widgets.clubCount(followed.length),
      followPrompt: copy.widgets.followPrompt,
      noFixtures: copy.widgets.noFixtures,
      versus: copy.widgets.versus,
      away: copy.widgets.away,
      homeTag: copy.widgets.homeTag,
      awayTag: copy.widgets.awayTag,
    },
    entries,
  };
}

/** ⚠ Also `news.json`'s directory — see `features/news/snapshot.ts`. */
export const SNAPSHOT_DIR = 'widget';
const SNAPSHOT_NAME = 'snapshot.json';

/** Where `_debug/widgets` and `Snapshot.swift` both look. */
export function snapshotFile(): File | null {
  const container = groupContainer();
  if (!container) return null;
  return new File(container, SNAPSHOT_DIR, SNAPSHOT_NAME);
}

/**
 * Write a JSON contract file where an extension can read it.
 *
 * ⚠ **Temp file then rename.** `File.write` is not documented atomic, and a
 * widget that reads a half-written file gets a decode failure and falls all the
 * way back to its empty state. A rename inside one volume is atomic, so the
 * extension sees either the old file or the new one.
 *
 * ⚠ Shared with `features/news/snapshot.ts` (ADR 0061) — the NEWS widget has a
 * second file in the same directory with the same hazard. One rename routine,
 * not two.
 *
 * Returns false when there is no container (Android, or a simulator without the
 * group provisioned). That is a normal answer — see `groupContainer`.
 */
export function writeGroupJson(directoryName: string, fileName: string, data: unknown): boolean {
  const container = groupContainer();
  if (!container) return false;

  try {
    const directory = new Directory(container, directoryName);
    if (!directory.exists) directory.create({ intermediates: true, idempotent: true });

    const temp = new File(directory, `${fileName}.tmp`);
    if (temp.exists) temp.delete();
    temp.create();
    temp.write(JSON.stringify(data));

    // ⚠ `moveSync`, not `move`. The async form returns a promise this function
    // does not await, which would put the rename in a race with the widget
    // reload fired immediately after it.
    temp.moveSync(new File(directory, fileName), { overwrite: true });
    return true;
  } catch {
    // A file that will not write leaves the previous one in place, which is
    // the better of the two failures.
    return false;
  }
}

/** Write the fixture snapshot where `Snapshot.swift` reads it. */
export function writeSnapshot(snapshot: WidgetSnapshot): boolean {
  return writeGroupJson(SNAPSHOT_DIR, SNAPSHOT_NAME, snapshot);
}

/**
 * The comparison key for "has anything actually changed".
 *
 * ⚠⚠ **`writtenAt` is excluded, and that is the entire point.** WidgetKit
 * budgets timeline reloads — roughly 40–70 a day for an actively used widget —
 * and spending them buys nothing but leaves the widget FROZEN for the rest of
 * the day with no error in any log. The re-arm runs on every foreground, so
 * comparing the timestamp would call every one of them a change.
 */
export function snapshotKey(snapshot: WidgetSnapshot): string {
  return JSON.stringify({ ...snapshot, writtenAt: '' });
}
